### PR TITLE
Add android auto support

### PR DESCRIPTION
**Pull request by Spacetech**
This change adds basic android auto support. A video of it in action for reference: https://youtu.be/Zl-ETVcv5BQ

Main changes:
- Implements a `MediaBrowserServiceCompat` which lets Android auto detect Jellyfin as a music player
- Supports library selection and drilling down into the library by different views (latest, albums, artists, songs, genres)
- Supports shuffling in the album view
- Supports prev / next tracks
- Shows thumbnails when possible

Notes:
- The `MediaService` (implementation of `MediaBrowserServiceCompat`) is a fairly standalone thing. Due to how android auto music players work, it will not rely on the main webview for ui. Instead it will use the `ApiClient` to load data from Jellyfin.
- Since it's a separate service and could/will launch independently from the main app, I added preferences for the server url, user id, and access token. Those preferences will be updated when the main app is launched. This makes it so the last used server will be used in android auto. Note: Saving the access token should be safe since the preferences are marked as private. 
- I named the folder `car` because that's more inline with what Google calls it - [Android for Cars](https://developer.android.com/training/cars)
- I based the url I use to load the audio track in the `createMediaSource` method on what the web client did. Is there a better way to obtain this url? (is there an api property for it?)

I see this as basic support that should be good enough to merge for now. There's certainly room for improvement. Here's a few ideas:
- Add shuffle button to more views
- Most of the lists are currently limited to 100 items. We need to figure out how to display more things in the android auto music ui. Some other music player apps list all the artist initials first and you slowly drill down to find what you're looking for. This has the advantage of supporting a large number of artists/albums. The disadvantage is that it requires a lot more user interaction to pick stuff out
- Better integration with the main app - Something like auto resuming whatever was last played would be great
- Add search support - Because clicking through large libraries to find a specific item is not fun
- Support for some sort of offline mode

This was my first time using Kotlin so let me know if I did anything weird. Thanks

References:
- Building media apps for cars - https://developer.android.com/training/cars/media
- uamp sample media app - https://github.com/android/uamp

### CODE REVIEW

**Comment by nielsvanvelzen** [comment]
Before I start reviewing I'd like to point out a few things:
- We will not include this in the initial 2.0.0 release. (But probably 2.1 as "main feature")
- We need to look into the publishing guidelines for Android Auto, I know Google is quite strict on those
- Since none of our Android contributors have Android Auto we won't be able to test it on real hardware, I think the emulator should work sufficiently though
This means that the PR might be open a while after approval before we merge it.

And lastly, thanks for this contribution! I'm really excited to see our efforts for the app rewrite are enabling features like this.

**Comment by neopc10** [comment]
Cool! Does this have playlist support?

**Comment by Spacetech** [comment]
@nielsvanvelzen That all sounds good. For reference, I have been doing the majority of testing on my phone using the [Android Auto for phones app](https://play.google.com/store/apps/details?id=com.google.android.projection.gearhead.phonescreen&hl=en_US)


@neopc10 Good idea. I just added that.

Going forward I don't want to add too much more to this pr, otherwise it may become harder to review.

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:164]:
```suggestion
                                    val itemImageUrl = apiClient.GetImageUrl(item, ImageOptions().apply {
                                        maxWidth = 180
                                        quality = 90
                                    })
```

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:168]: Only pass the item id as media id, if you need the library id you can pass that in a bundle using `setExtras`

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:148]: Please add an extension function in `utils/ApiExtensions.kt` to use coroutines instead of callbacks for the GetUserViews function.

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:168]: (Same for the other places that used this way of communicating data)

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:148]: (Same for other used api functions)

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:295]:
```suggestion
                        query.includeItemTypes = arrayOf(BaseItemType.MusicAlbum)
```

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:295]: (also in some other places)

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:289]:
```suggestion
                        query.sortBy = arrayOf(ItemSortBy.SortName)
```

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:289]: (also in some other places)

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:586]: I'm quite sure we have a function to generate this URL but can't find it right now.

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

**Comment by nielsvanvelzen** [comment]
I had some issues to get the Android Auto app working but fortunately Google made the "Android Media Controller Test" app which is useful for testing the media browser:

https://github.com/googlesamples/android-media-controller

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (Maxr1998) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:148]: > Please add an extension function in `utils/ApiExtensions.kt` to use coroutines instead of callbacks for the GetUserViews function.

Wrt this: I think we shouldn't merge this before we added the new api client in, so maybe it makes sense to wait a bit when it has to be updated/changed anyways at a later point.

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:148]: I'm not sure when the new apiclient is ready so I'd rather merge this when it's ready and 2.0.0 is released instead of waiting for the apiclient.

**Comment by mholt** [comment]
Sorry I am of no help here whatsoever, but I just want to cheer you on from the sidelines, sooooo yeah! :partying_face:  :raised_hands:  Really looking forward to this!

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (Spacetech) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:168]: Unfortunately I don't think this could work. The `onLoadChildren` method only has that `parentId` string arg. This means the media id has to have all the information required to load the next view, which is why I'm playing with strings.

There is an overload method of [onLoadChildren with a bundle](https://developer.android.com/reference/androidx/media/MediaBrowserServiceCompat#onLoadChildren(java.lang.String,%20androidx.media.MediaBrowserServiceCompat.Result%3Cjava.util.List%3Candroid.support.v4.media.MediaBrowserCompat.MediaItem%3E%3E,%20android.os.Bundle)), but I never saw it called while testing. There's a [stackoverflow ](https://stackoverflow.com/questions/55863290/in-mediabrowserservice-onloadchildren-when-how-is-the-options-bundle-supplied)post about it but it has no answer.

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (Spacetech) [reply app/src/main/java/org/jellyfin/mobile/car/MediaService.kt]: I added the api extensions. Since `onLoadChildren` is not a suspend function, I had to add a `CoroutineScope` to media service. Let me know if there's a better way to do that. It seems to be working well so it's probably okay.

File: buildSrc/src/main/kotlin/Dependencies.kt
Comment (nielsvanvelzen) [inline buildSrc/src/main/kotlin/Dependencies.kt:98]: Shouldn't call this "car" since the media library / mediabrowser feature is not specifically for cars.

File: buildSrc/src/main/kotlin/Dependencies.kt
Comment (nielsvanvelzen) [inline buildSrc/src/main/kotlin/Dependencies.kt:100]: Version isn't aligned with the actual ExoPlayer extension, should use `Versions.exoPlayer`.

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:1]: Shouldn't call this "car" since the media library / mediabrowser feature is not specifically for cars.

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:592]: Move all constants to a companion object in MediaService

File: app/src/main/java/org/jellyfin/mobile/car/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/car/MediaService.kt:163]: Use the enum here

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: app/build.gradle.kts
Comment (nielsvanvelzen) [inline app/build.gradle.kts:90]: This comment still says car

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]
Looks good! Just one tiny change left.

File: app/src/main/java/org/jellyfin/mobile/AppPreferences.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/AppPreferences.kt:36]: I'm not sure whether I want to save those into SharedPreferences.. I've already integrated a database for server management in the `compose` branch, and I think I'll cherry-pick that very soon. If it's ok for you, I'd wait for that and refactor your credential persistence to use the new database then.

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:97]: Consider using `apply` here.

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:107]:
```suggestion
```

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:105]:
```suggestion
```

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:121]:
```suggestion
```

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:218]:
```suggestion
                    val views = libraryViews.map { item ->
```

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:573]: Can you break this up or potentially (don't know if currently possible) use the API client for this?

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:547]:
```suggestion
```

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:557]:
```suggestion
        ) {}
```

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:459]:
```suggestion
```

**Review by Maxr1998 — CHANGES_REQUESTED** [review]
Thanks for the contribution! The changes generally look quite good, I'm not happy with some smaller parts and some of the formatting however. Also, I think some things require advance changes to the rest of the app, as described in the individual comments.

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Spacetech) [reply app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:573]: Niel mentioned that too but we don't see an api client method for it. How should I break it up?

File: app/src/main/java/org/jellyfin/mobile/AppPreferences.kt
Comment (Spacetech) [reply app/src/main/java/org/jellyfin/mobile/AppPreferences.kt:36]: Sounds good to me. Let me know when that's available.

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (Maxr1998) [reply app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:565]: You could split it into multiple lines for example. Also, I recommend using the Kotlin format strings, like `"string${variable}string"`.

Comment (Spacetech) [reply app/src/main/java/org/jellyfin/mobile/media/MediaService.kt]: Thanks. Looks much better now

**Comment by Maxr1998** [comment]
@Spacetech I've decided to not do the database stuff and merge this pretty much as is. I rebased it onto the latest master and fixed the conflicts and squashed the commits in that process. If it's ok for you, I'd force push the changes to this branch. Then, both Niels and me will give it another review, so we can hopefully merge it!

**Comment by Spacetech** [comment]
@Maxr1998 Yea that sounds great. Go for it

**Comment by Maxr1998** [comment]
Perfect! I just noticed that since we update ExoPlayer meanwhile, this uses deprecated APIs now. I'll fix those in a minute :grin:

**Review by Maxr1998 — APPROVED** [review]

**Comment by Maxr1998** [comment]
Updated to the latest ExoPlayer APIs, refactored some code, fixed the album covers for the currently playing song (didn't work for me, cause only the albums have images in my config, now it falls back to those), and gave it a final test, I'm happy now :blush:

File: app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/media/MediaService.kt:556]: Looking at https://github.com/jellyfin/jellyfin-apiclient-java/blob/master/jellyfin-api/src/main/kotlin-generated/org/jellyfin/apiclient/api/operations/UniversalAudioApi.kt#L43

- The deviceId should probably be plaintext, not base64
- PlaySessionId doesn't exist
- StartTimeTicks _should_ default to 0 when missing
- EnableRedirection is true by default

**Review by nielsvanvelzen — APPROVED** [review]

**Comment by dkanada** [comment]
@Spacetech if you have any interest in adding this to Gelli feel free :sweat_smile: lots of people ask about Android Auto but I can't exactly test it easily without a device that supports it.

**Comment by nielsvanvelzen** [comment]
Hey @Spacetech you mind it if I use your YouTube video in the 2.1 release notes?

**Comment by Spacetech** [comment]
@nielsvanvelzen I don't mind, go for it.

**Pull request merged**
